File: web/client/src/workers/sqlglot/sqlglot.py
import json

import sqlglot
from sqlglot import exp
from sqlglot.dialects.dialect import Dialect, DialectType

dialects = json.dumps(
    [
        {
            "dialect_title": "SQLGlot" if dialect == "" else dialect_class.__name__,
            "dialect_name": dialect,
        }
        for dialect, dialect_class in Dialect.classes.items()
    ]
)


def parse_to_json(sql: str, read: DialectType = None) -> str:
    return json.dumps(
        [exp.dump() if exp else {} for exp in sqlglot.parse(sql, read=read, error_level="ignore")]
    )


def get_dialect(name: str = "") -> str:
    dialect = Dialect.classes.get(name, Dialect)
    keywords = dialect.tokenizer_class.KEYWORDS
    type_mapping = dialect.generator_class.TYPE_MAPPING

    return json.dumps(
        {
            "keywords": " ".join(keywords),
            "types": " ".join(type_mapping.get(t, t.value) for t in exp.DataType.Type),
        }
    )


def format(sql: str = "", read: DialectType = None) -> str:
    return "\n".join(
        sqlglot.transpile(sql, read=read, error_level=sqlglot.errors.ErrorLevel.IGNORE, pretty=True)
    )


def validate(sql: str = "", read: DialectType = None) -> str:
    try:
        sqlglot.transpile(
            sql, read=read, pretty=False, unsupported_level=sqlglot.errors.ErrorLevel.IMMEDIATE
        )
    except sqlglot.errors.ParseError as e:
        return json.dumps(False)

    return json.dumps(True)


{
    "dialects": dialects,
    "get_dialect": get_dialect,
    "parse_to_json": parse_to_json,
    "validate": validate,
    "format": format,
}

File: web/client/src/workers/sqlglot/sqlglot.ts
declare function importScripts(...urls: string[]): void

importScripts('https://cdn.jsdelivr.net/pyodide/v0.23.2/full/pyodide.js')

const scope = self as any

async function loadPyodideAndPackages(): Promise<{
  sqlglot: Record<string, any>
}> {
  scope.pyodide = await scope.loadPyodide()

  await scope.pyodide.loadPackage('micropip')

  const micropip = scope.pyodide.pyimport('micropip')

  await micropip.install('sqlglot')
  await micropip.install('typing-extensions')

  const file = await (
    await fetch(new URL('./sqlglot.py', import.meta.url))
  ).text()

  scope.postMessage({
    topic: 'init',
  })

  return {
    sqlglot: scope.pyodide.runPython(file),
  }
}

const pyodideReadyPromise = loadPyodideAndPackages()

scope.onmessage = async (e: MessageEvent) => {
  const { sqlglot } = await pyodideReadyPromise

  if (e.data.topic === 'validate') {
    let payload: boolean

    try {
      payload = JSON.parse(sqlglot.get('validate')?.(e.data.payload))
    } catch (error) {
      payload = false
    }

    scope.postMessage({
      topic: 'validate',
      payload,
    })
  }

  if (e.data.topic === 'dialect') {
    let payload: { keywords: string; types: string }

    try {
      payload = JSON.parse(sqlglot.get('get_dialect')?.(e.data.payload))
    } catch (error) {
      payload = {
        keywords: '',
        types: '',
      }
    }

    scope.postMessage({
      topic: 'dialect',
      payload,
    })
  }

  if (e.data.topic === 'dialects') {
    let payload: Array<{ dialect_title: string; dialect_name: string }>

    try {
      payload = JSON.parse(sqlglot.get('dialects'))
    } catch (error) {
      payload = []
    }

    scope.postMessage({
      topic: 'dialects',
      payload,
    })
  }

  if (e.data.topic === 'format') {
    scope.postMessage({
      topic: 'format',
      payload: sqlglot.get('format')?.(e.data.payload.sql) ?? '',
    })
  }

  if (e.data.topic === 'init') {
    scope.postMessage({
      topic: 'init',
    })
  }
}
